client/src/main/java/io/lionweb/client/inmemory/RepositoryData.java

```diff
@@ -139,8 +172,115 @@ void retrieve(String nodeId, int limit, List<SerializedClassifierInstance> retri
     }
     retrieved.add(node);
     if (limit > 0) {
-      node.getChildren().forEach(c -> retrieve(c, limit - 1, retrieved));
-      node.getAnnotations().forEach(a -> retrieve(a, limit - 1, retrieved));
+      node.getChildren()
+          .forEach(
+              c -> {
+                try {
+                  retrieve(c, limit - 1, retrieved);
+                } catch (Exception e) {
+                  throw new RuntimeException("Unable to retrieve child of " + node, e);
+                }
+              });
+      node.getAnnotations()
+          .forEach(
+              a -> {
+                try {
+                  retrieve(a, limit - 1, retrieved);
+                } catch (Exception e) {
+                  throw new RuntimeException("Unable to retrieve annotation of " + node, e);
+                }
+              });
     }
   }
+
+  /** This is intended for debugging purposes. It checks if the data is consistent. */
+  public @NotNull ValidationResult checkConsistency() {
+    ValidationResult result = new ValidationResult();
+
+    // Check for invalid node IDs
+    for (String nodeId : nodesByID.keySet()) {
+      if (!CommonChecks.isValidID(nodeId)) {
+        result.addError("Invalid node id: " + nodeId);
+      }
+    }
+
+    // Check for duplicate node IDs (this is already guaranteed by using HashMap, but good to be
+    // explicit)
+    // No need for additional check since HashMap ensures uniqueness
+
+    // Ensuring that containments and annotations are the inverse of parent relationships
+    Map<String, Set<String>> containedNodes = new HashMap<>();
+    for (SerializedClassifierInstance node : nodesByID.values()) {
+      for (SerializedContainmentValue containmentValue : node.getContainments()) {
+        for (String childId : containmentValue.getChildrenIds()) {
+          // Verifying nodes do not appear in multiple containments or annotations
+          String newPlacement = node.getID() + " at " + containmentValue.getMetaPointer();
+          if (containedNodes.containsKey(childId)) {
+            result.addError(
+                childId
+                    + " is listed in multiple places: "
+                    + containedNodes.get(childId)
+                    + " and now "
+                    + newPlacement);
+          } else {
+            containedNodes.put(childId, new HashSet<>(Arrays.asList(newPlacement)));
+          }
+          SerializedClassifierInstance child = nodesByID.get(childId);
+          if (child != null && !child.getParentNodeID().equals(node.getID())) {
+            result.addError(
+                childId
+                    + " is listed as child of "
+                    + node.getID()
+                    + " but it has as parent "
```

Comment:
Can just as well phrase it as `" but it has ${parentID} as parent"`.

Reply:
Changed as suggested